File: python/백준/구현/14503 로봇 청소기.py
import sys

input = sys.stdin.readline

N, M = map(int, input().split())
r, c, d = list(map(int, input().split()))
room = [list(map(int, input().split())) for _ in range(N)]
directions = [(-1, 0), (0, 1), (1, 0), (0, -1)]

def left(d):
    return (d + 3) % 4

count = 0

while True:
    if room[r][c] == 0:
        room[r][c] = 2
        count += 1
        
    moved = False

    for _ in range(4):
        d = left(d)
        dr, dc = directions[d]
        nr, nc = r + dr, c + dc
        if 0 <= nc < M and 0 <= nr < N and room[nr][nc] == 0:
            r, c = nr, nc
            moved = True
            break
            
    if moved:
        continue

    back_dir = (d + 2) % 4
    br, bc = directions[back_dir]
    nr, nc = r + br, c + bc

    if 0 <= nr < N and 0 <= nc < M and room[nr][nc] != 1:
        r, c = nr, nc
    else:
        print(count)
        break

File: cpp/백준/구현/14503 로봇 청소기.cpp
#include <iostream>
#include <vector>
using namespace std;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    int N, M;
    cin >> N >> M;

    int r, c, d;
    cin >> r >> c >> d;

    vector<vector<int>> room(N, vector<int>(M));
    for (int i = 0; i < N; ++i)
        for (int j = 0; j < M; ++j)
            cin >> room[i][j];

    int dr[] = {-1, 0, 1, 0};
    int dc[] = {0, 1, 0, -1};

    int count = 0;
    while (true) {
        if (room[r][c] == 0) {
            room[r][c] = 2;
            ++count;
        }

        bool moved = false;
        for (int i = 0; i < 4; ++i) {
            d = (d + 3) % 4;
            int nr = r + dr[d];
            int nc = c + dc[d];
            if (nr >= 0 && nr < N && nc >= 0 && nc < M && room[nr][nc] == 0) {
                r = nr;
                c = nc;
                moved = true;
                break;
            }
        }

        if (moved) continue;

        int backDir = (d + 2) % 4;
        int nr = r + dr[backDir];
        int nc = c + dc[backDir];

        if (nr >= 0 && nr < N && nc >= 0 && nc < M && room[nr][nc] != 1) {
            r = nr;
            c = nc;
        } else {
            cout << count << "\n";
            break;
        }
    }

    return 0;
}
